<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>Document</title>
</head>
<body>
	<script type="text/javascript">
	var cat = {
		name: 'aa'
	}
	function beatTheMonster(a, b, c) {
		console.log(this.name)
		console.log('a: ',a)
		console.log('b: ',b)
		console.log('c: ',c)
	}
	Function.prototype.myCall = function(context) {
		console.log('this; ',this)
		console.log('context: ',context)
		console.log('arguments: ',arguments)
		// 默认执行window
		context = context || window
		var args = [];
		// cat这个对象添加beat方法
		context.fn = this;
		// ES3
		// for (var i = 1,len = arguments.length; i < len; i++) {
		// 	args.push('arguments['+ i+']');
		// };
		// console.log('args: ',args)
		// eval('context.fn('+args+')');
		// ES6
		for (var i = 1; i < arguments.length; i++) {
			args.push(arguments[i])
		};
		// 说好的剩余运算符呢，怎么都出来了，看来并不是剩余参数，而是全部
		console.log('...args: ',...args)
		// 为什么只打印了第一个，因为它确实只能执行一次，要打印多个，待加实参
		// console.log('args: ',args)
		var result  = context.fn(...args);
		delete context.fn
		return result

	};
	beatTheMonster.myCall(cat, 'xxx', 10);
	var name = 'windowName'
	console.log('return: ',beatTheMonster.myCall(null))
	// 目的：1、测试call(添加，执行，删除)；2、测试call传参，第二个参数后面
    // 关键点： 1、如果传入对象为null;2、函数通常会有返回值
	</script>
</body>
</html>